src/disk_io_thread_pool.cpp

```diff
@@ -0,0 +1,200 @@
+CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+POSSIBILITY OF SUCH DAMAGE.
+
+*/
+
+#include "libtorrent/disk_io_thread_pool.hpp"
+#include "libtorrent/disk_io_thread.hpp"
+#include <functional>
+
+namespace
+{
+	constexpr std::chrono::seconds reap_idle_threads_interval(60);
+}
+
+namespace libtorrent
+{
+	disk_io_thread_pool::disk_io_thread_pool(pool_thread_interface& thread_iface
+		, io_service& ios)
+		: m_thread_iface(thread_iface)
+		, m_max_threads(0)
+		, m_threads_to_exit(0)
+		, m_abort(false)
+		, m_num_idle_threads(0)
+		, m_min_idle_threads(0)
+		, m_idle_timer(ios)
+	{}
+
+	disk_io_thread_pool::~disk_io_thread_pool()
+	{
+		abort(true);
+	}
+
+	void disk_io_thread_pool::set_max_threads(int const i)
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (i == m_max_threads) return;
+		m_max_threads = i;
+		if (m_threads.size() < i) return;
+		stop_threads(int(m_threads.size()) - i);
+	}
+
+	void disk_io_thread_pool::abort(bool wait)
+	{
+		std::unique_lock<std::mutex> l(m_mutex);
+		if (m_abort) return;
+		m_max_threads = 0;
+		m_abort = true;
+		m_idle_timer.cancel();
+		stop_threads(int(m_threads.size()));
+		for (auto& t : m_threads)
+		{
+			if (wait)
+			{
+				// must release m_mutex to avoid a deadlock if the thread
+				// tries to aquire it
+				l.unlock();
+				t.join();
+				l.lock();
+			}
+			else
+				t.detach();
+		}
+		m_threads.clear();
+	}
+
+	void disk_io_thread_pool::thread_active()
+	{
+		--m_num_idle_threads;
+		TORRENT_ASSERT(m_num_idle_threads >= 0);
+
+		int current_min = m_min_idle_threads;
+		while (m_num_idle_threads < current_min
+			&& !m_min_idle_threads.compare_exchange_weak(current_min, m_num_idle_threads));
+	}
+
+	bool disk_io_thread_pool::try_thread_exit(std::thread::id id)
+	{
+		int to_exit = m_threads_to_exit;
+		while (to_exit > 0 &&
+			!m_threads_to_exit.compare_exchange_weak(to_exit, to_exit - 1));
+		if (to_exit > 0)
+		{
+			std::unique_lock<std::mutex> l(m_mutex);
+			if (!m_abort)
+			{
+				auto new_end = std::remove_if(m_threads.begin(), m_threads.end()
+					, [id, &l](std::thread& t)
+				{
+					if (t.get_id() == id)
+					{
+						t.detach();
+						return true;
+					}
+					return false;
+				});
+				TORRENT_ASSERT(new_end != m_threads.end());
+				m_threads.erase(new_end, m_threads.end());
+				if (m_threads.empty()) m_idle_timer.cancel();
+			}
+		}
+		return to_exit > 0;
+	}
+
+	std::thread::id disk_io_thread_pool::first_thread_id()
+	{
+		std::lock_guard<std::mutex> l(m_mutex);
+		if (m_threads.empty()) return std::thread::id();
+		return m_threads.front().get_id();
+	}
+
+	void disk_io_thread_pool::job_queued(int queue_size)
```

Comment:
when is this called? the name "job_queued" sounds like it would return the size of the job queue, but it isn't doing that.
is it informing the thread pool of how many jobs are queued? and it's called "job_queued" because that's the event it's called on, another job was just queued?
I think a comment here may make sense too